Add Web Extension interfaces for windows and tabs with stub implementations.

xeenon opened the pull request:
#### aebc1969539a8aeec18bc36f907f3b48c8f13cd7
<pre>
Add Web Extension interfaces for windows and tabs with stub implementations.
<a href="https://webkit.org/b/260664">https://webkit.org/b/260664</a>
rdar://problem/114387424

Reviewed by Brian Weinstein.

Implementation to come in future pull requests.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::alarms):
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::permissions):
(WebKit::WebExtensionAPINamespace::runtime):
(WebKit::WebExtensionAPINamespace::tabs):
(WebKit::WebExtensionAPINamespace::test):
(WebKit::WebExtensionAPINamespace::windows):
(WebKit::WebExtensionAPINamespace::webNavigation):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm: Added.
(WebKit::WebExtensionAPITabs::isPropertyAllowed):
(WebKit::WebExtensionAPITabs::createTab):
(WebKit::WebExtensionAPITabs::query):
(WebKit::WebExtensionAPITabs::get):
(WebKit::WebExtensionAPITabs::getCurrent):
(WebKit::WebExtensionAPITabs::getSelected):
(WebKit::WebExtensionAPITabs::duplicate):
(WebKit::WebExtensionAPITabs::update):
(WebKit::WebExtensionAPITabs::remove):
(WebKit::WebExtensionAPITabs::reload):
(WebKit::WebExtensionAPITabs::goBack):
(WebKit::WebExtensionAPITabs::goForward):
(WebKit::WebExtensionAPITabs::getZoom):
(WebKit::WebExtensionAPITabs::setZoom):
(WebKit::WebExtensionAPITabs::detectLanguage):
(WebKit::WebExtensionAPITabs::toggleReaderMode):
(WebKit::WebExtensionAPITabs::onActivated):
(WebKit::WebExtensionAPITabs::onAttached):
(WebKit::WebExtensionAPITabs::onCreated):
(WebKit::WebExtensionAPITabs::onDetached):
(WebKit::WebExtensionAPITabs::onHighlighted):
(WebKit::WebExtensionAPITabs::onMoved):
(WebKit::WebExtensionAPITabs::onRemoved):
(WebKit::WebExtensionAPITabs::onReplaced):
(WebKit::WebExtensionAPITabs::onUpdated):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIWebNavigationEvent::addListener):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm: Added.
(WebKit::WebExtensionAPIWindows::isPropertyAllowed):
(WebKit::WebExtensionAPIWindows::createWindow):
(WebKit::WebExtensionAPIWindows::get):
(WebKit::WebExtensionAPIWindows::getCurrent):
(WebKit::WebExtensionAPIWindows::getLastFocused):
(WebKit::WebExtensionAPIWindows::getAll):
(WebKit::WebExtensionAPIWindows::update):
(WebKit::WebExtensionAPIWindows::remove):
(WebKit::WebExtensionAPIWindows::onCreated):
(WebKit::WebExtensionAPIWindows::onRemoved):
(WebKit::WebExtensionAPIWindows::onFocusChanged):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm: Added.
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIWindowsEvent::addListener):
(WebKit::WebExtensionAPIWindowsEvent::removeListener):
(WebKit::WebExtensionAPIWindowsEvent::hasListener):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIEvent.h:
(WebKit::WebExtensionAPIEvent::WebExtensionAPIEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h: Added.
(WebKit::WebExtensionAPITabs::tabIdentifierNone const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebNavigationEvent.h:
(WebKit::WebExtensionAPIWebNavigationEvent::WebExtensionAPIWebNavigationEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindows.h: Added.
(WebKit::WebExtensionAPIWindows::windowIdentifierNone const):
(WebKit::WebExtensionAPIWindows::windowIdentifierCurrent const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindowsEvent.h: Added.
(WebKit::WebExtensionAPIWindowsEvent::listeners const):
(WebKit::WebExtensionAPIWindowsEvent::WebExtensionAPIWindowsEvent):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindows.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindowsEvent.idl: Added.

Canonical link: <a href="https://commits.webkit.org/267238@main">https://commits.webkit.org/267238@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33df1ea3d1d655935e60decaff142e2648714162

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17819 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15074 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19386 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16479 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16248 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16711 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18582 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13954 "Passed tests") | | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21380 "Failed to checkout and rebase branch from PR 17022") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14943 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14681 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/17929 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15282 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12957 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14516 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18887 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1967 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15102 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/DerivedSources-input.xcfilelist:
```diff
@@ -375,9 +375,12 @@ $(PROJECT_DIR)/WebProcess/Extensions/Interfaces/WebExtensionAPIExtension.idl
 $(PROJECT_DIR)/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl
 $(PROJECT_DIR)/WebProcess/Extensions/Interfaces/WebExtensionAPIPermissions.idl
 $(PROJECT_DIR)/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl
+$(PROJECT_DIR)/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl
```
Will there be a TabsEvent.idl as well?

xeenon replied:
No, only Windows has a special event since it has a filter. Tabs uses the generic event already in the code.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
```diff
@@ -65,7 +65,7 @@ class WebExtensionTab : public RefCounted<WebExtensionTab> {
         URL        = 1 << 8,
         ZoomFactor = 1 << 9,
-        All        = std::numeric_limits<uint16_t>::max(),
+        All        = Audible | Loading | Muted | Pinned | ReaderMode | Size | Title | URL | ZoomFactor,
```
This is very cool!

xeenon replied:
This is suppose to be on my other PR. I will move it there when I land.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
```diff
@@ -0,0 +1,118 @@
+
+namespace WebKit {
+
+void WebExtensionAPIWindowsEvent::invokeListenersWithArgument(id argument, WindowType windowType)
```
Oh, that's right, the WIndowsEvent exists because of the WindowType filter.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
```diff
@@ -0,0 +1,118 @@
+            windowTypeFilter.add(WindowType::Popup);
+
+        if (!windowTypeFilter) {
+            *outExceptionString = @"The 'windowTypes' array must contain 'normal' and/or 'popup'.";
```
If a developer specified both `normal` and `popup` - would that be the same as not having a filter at all?

xeenon replied:
Yeah, they both end up as `WindowTypes::All` in that case.

xeenon committed:
Add Web Extension interfaces for windows and tabs with stub implementations.
https://webkit.org/b/260664
rdar://problem/114387424

Reviewed by Brian Weinstein.

Implementation to come in future pull requests.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::alarms):
(WebKit::WebExtensionAPINamespace::extension):
(WebKit::WebExtensionAPINamespace::permissions):
(WebKit::WebExtensionAPINamespace::runtime):
(WebKit::WebExtensionAPINamespace::tabs):
(WebKit::WebExtensionAPINamespace::test):
(WebKit::WebExtensionAPINamespace::windows):
(WebKit::WebExtensionAPINamespace::webNavigation):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm: Added.
(WebKit::WebExtensionAPITabs::isPropertyAllowed):
(WebKit::WebExtensionAPITabs::createTab):
(WebKit::WebExtensionAPITabs::query):
(WebKit::WebExtensionAPITabs::get):
(WebKit::WebExtensionAPITabs::getCurrent):
(WebKit::WebExtensionAPITabs::getSelected):
(WebKit::WebExtensionAPITabs::duplicate):
(WebKit::WebExtensionAPITabs::update):
(WebKit::WebExtensionAPITabs::remove):
(WebKit::WebExtensionAPITabs::reload):
(WebKit::WebExtensionAPITabs::goBack):
(WebKit::WebExtensionAPITabs::goForward):
(WebKit::WebExtensionAPITabs::getZoom):
(WebKit::WebExtensionAPITabs::setZoom):
(WebKit::WebExtensionAPITabs::detectLanguage):
(WebKit::WebExtensionAPITabs::toggleReaderMode):
(WebKit::WebExtensionAPITabs::onActivated):
(WebKit::WebExtensionAPITabs::onAttached):
(WebKit::WebExtensionAPITabs::onCreated):
(WebKit::WebExtensionAPITabs::onDetached):
(WebKit::WebExtensionAPITabs::onHighlighted):
(WebKit::WebExtensionAPITabs::onMoved):
(WebKit::WebExtensionAPITabs::onRemoved):
(WebKit::WebExtensionAPITabs::onReplaced):
(WebKit::WebExtensionAPITabs::onUpdated):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIWebNavigationEvent::addListener):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm: Added.
(WebKit::WebExtensionAPIWindows::isPropertyAllowed):
(WebKit::WebExtensionAPIWindows::createWindow):
(WebKit::WebExtensionAPIWindows::get):
(WebKit::WebExtensionAPIWindows::getCurrent):
(WebKit::WebExtensionAPIWindows::getLastFocused):
(WebKit::WebExtensionAPIWindows::getAll):
(WebKit::WebExtensionAPIWindows::update):
(WebKit::WebExtensionAPIWindows::remove):
(WebKit::WebExtensionAPIWindows::onCreated):
(WebKit::WebExtensionAPIWindows::onRemoved):
(WebKit::WebExtensionAPIWindows::onFocusChanged):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm: Added.
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIWindowsEvent::addListener):
(WebKit::WebExtensionAPIWindowsEvent::removeListener):
(WebKit::WebExtensionAPIWindowsEvent::hasListener):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIEvent.h:
(WebKit::WebExtensionAPIEvent::WebExtensionAPIEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h: Added.
(WebKit::WebExtensionAPITabs::tabIdentifierNone const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebNavigationEvent.h:
(WebKit::WebExtensionAPIWebNavigationEvent::WebExtensionAPIWebNavigationEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindows.h: Added.
(WebKit::WebExtensionAPIWindows::windowIdentifierNone const):
(WebKit::WebExtensionAPIWindows::windowIdentifierCurrent const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindowsEvent.h: Added.
(WebKit::WebExtensionAPIWindowsEvent::listeners const):
(WebKit::WebExtensionAPIWindowsEvent::WebExtensionAPIWindowsEvent):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindows.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindowsEvent.idl: Added.

Canonical link: https://commits.webkit.org/267238@main
